File: pe15/pe15.py
# Moving only left and right, how many paths exist from the top-left corner to the bottom-right corner in a 20x20 grid?

import math

def paths(r=20, c=20):
  f = math.factorial
  return f((r) + (c)) / (f(r) * f(c))

def test():
  assert(paths(2, 2) == 6)
  assert(paths(1,1) == 2)


test()

def solve():
  print("Number of paths in 20x20 grid = %d" % paths())

solve()

File: pe15/pe15.go
package main

import (
	"fmt"
	"math/big"
)

func paths(r, c *big.Int) *big.Int {
	num, denom, res := new(big.Int), new(big.Int), new(big.Int)
	num.Add(r, c)
	num.Set(factorial(num))
	denom.Mul(factorial(r), factorial(c))
	res.Div(num, denom)
	return res
}

func factorial(n *big.Int) *big.Int {
	result := big.NewInt(1)
	for i := big.NewInt(2); i.Cmp(n) <= 0; i.Add(i, big.NewInt(1)) {
		result.Mul(result, i)
	}
	return result
}

func main() {
	r, c := big.NewInt(int64(20)), big.NewInt(int64(20))
	fmt.Printf("The number of paths that exist from the top-left corner to the bottom-right corner in a 20x20 grid is %s\n", paths(r, c))
}
